File: single-source-shortest-paths/python/dijkstra.py
import heapq

from graph import Edge, Graph, Vertex, initialize_single_source, relax


def dijkstra(graph, s):
    initialize_single_source(graph, s)

    vertices = list(graph.vertices.values())

    heapq.heapify(vertices)

    adjacency_list = graph.adjacency_list
    while vertices:
        u = heapq.heappop(vertices)
        adjacencies = adjacency_list[u.name]
        for v, weight in adjacencies:
            relax(u, v, weight)

            heapq.heapify(vertices)


def print_path(s, v, v_name):
    if v == s:
        print(s.name, end=" -> " if v.name != v_name else "")
    elif v.predecessor:
        print_path(s, v.predecessor, v_name)
        print(v.name, end=" -> " if v.name != v_name else "")


if __name__ == "__main__":
    vertices = {}
    for name in "styxz":
        vertices[name] = Vertex(name)

    edges = []
    edges.append(Edge("s", "t", 10))
    edges.append(Edge("s", "y", 5))
    edges.append(Edge("t", "y", 2))
    edges.append(Edge("t", "x", 1))
    edges.append(Edge("y", "t", 3))
    edges.append(Edge("y", "x", 9))
    edges.append(Edge("y", "z", 2))
    edges.append(Edge("x", "z", 4))
    edges.append(Edge("z", "s", 7))
    edges.append(Edge("z", "x", 6))

    graph = Graph(vertices, edges)

    print("--- graph ---")
    graph.print_graph()

    print("--- Dikstra's algorithm ---")
    dijkstra(graph, vertices["s"])

    print("--- result ---")
    for v in vertices.values():
        print(f"{v.name} : {v.distance} ", end="(")
        print_path(vertices["s"], v, v.name)
        print(")")

File: single-source-shortest-paths/golang/dijkstra.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

func dijkstra(g *Graph, s *Vertex) {
	vertexHeap := make(vertexHeap, 0)
	vertices := g.vertices

	for _, v := range vertices {
		vertexHeap = append(vertexHeap, v)
	}

	for _, v := range vertexHeap {
		if v.name != s.name {
			v.distance = math.MaxInt64
		} else {
			v.distance = 0
		}
	}

	heap.Init(&vertexHeap)

	adjacencyList := g.adjacencyList
	for len(vertexHeap) > 0 {
		u := heap.Pop(&vertexHeap).(*Vertex)
		adjacencies := adjacencyList[u.name]
		for _, adjacency := range adjacencies {
			v := vertices[adjacency.end]
			weight := adjacency.weight
			relax(u, v, weight)

			heap.Init(&vertexHeap)
		}
	}
}

func printReseult(s *Vertex, v *Vertex, v_name string) {
	if v == s {
		fmt.Printf("%s", s.name)
		if v.name != v_name {
			fmt.Print(" -> ")
		}
	} else if v.predecessor != nil {
		printReseult(s, v.predecessor, v_name)
		fmt.Printf("%s", v.name)
		if v.name != v_name {
			fmt.Print(" -> ")
		}
	}
}

func main() {
	vertices := make(map[string]*Vertex)
	for _, name := range "styxz" {
		vertices[string(name)] = &Vertex{name: string(name)}
	}

	edges := make([]*Edge, 0)
	edges = append(edges, &Edge{start: "s", end: "t", weight: 10})
	edges = append(edges, &Edge{start: "s", end: "y", weight: 5})
	edges = append(edges, &Edge{start: "t", end: "y", weight: 2})
	edges = append(edges, &Edge{start: "t", end: "x", weight: 1})
	edges = append(edges, &Edge{start: "y", end: "t", weight: 3})
	edges = append(edges, &Edge{start: "y", end: "x", weight: 9})
	edges = append(edges, &Edge{start: "y", end: "z", weight: 2})
	edges = append(edges, &Edge{start: "x", end: "z", weight: 4})
	edges = append(edges, &Edge{start: "z", end: "s", weight: 7})
	edges = append(edges, &Edge{start: "z", end: "x", weight: 6})

	graph := &Graph{vertices: vertices, edges: edges}
	graph.initializeGraph()

	fmt.Println("--- graph ---")
	graph.printGraph()

	fmt.Println("--- Dijkstra's algorithm ---")
	dijkstra(graph, vertices["s"])

	fmt.Println("--- result ---")
	for _, v := range vertices {
		fmt.Printf("%s (%d): ", v.name, v.distance)
		printReseult(vertices["s"], v, v.name)
		fmt.Println()
	}
}
